### PR TITLE
Adding support for Razor precompilation

### LINKED ISSUE #3917: Razor Precompilation
#### Summary

We want to bring back Razor Precompilation so that you can ship an assembly or app with views baked in. We're going to start developing this during the 1.1.0 timeframe based on a prototype from https://github.com/aspnet/Performance/tree/dev/testapp/RazorCodeGenerator

**Goals:**
- Faster startup time (traded for build time)
- Find errors at build time
- Same Razor programming build as runtime compilation
- Loading/detection of precompiled views on by default (doing compilation requires a gesture)

**Non-Goals:**
- Support for replacing a precompiled view with a 'local' view

#### Details

Our plan is to implement precompilation via a command line tool. We have an existing tool that can generate `.cs` files given a project - we need to make that tool production quality and then add the ability to compile the generated code and embed it as a resource in the application assembly.

We'll also add support to the view engine to detect and use embedded view assemblies automatically via application parts. This means that dependencies you reference can include precompiled views.

The biggest unknown here is how exactly we'll interface with your application's startup code path when you've configured Razor to include extra application parts or references. We need to start productionizing what we have and look at gaps.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #3917

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/IDesignTimeMvcBuilderConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Mvc
+{
+    /// <summary>
+    /// Configures the <see cref="IMvcBuilder"/>. Implement this interface to enable design-time configuration
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Core/IDesignTimeMvcBuilderConfiguration.cs]: We may want to be more specific here so people don't think this enables custom TagHelper resolution in their cshtml editors.

File: src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ViewInfoContainer.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationParts
+{
+    /// <summary>
+    /// A container for <see cref="ViewInfo"/> instances.
+    /// </summary>
+    public class ViewInfoContainer
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ViewInfoContainer.cs]: Just double checking, this name change hasn't been updated in the other PR yet right?

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeature.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Compilation
+{
+    public class ViewsFeature
+    {
+        public IDictionary<string, Type> Views { get; } =
+            new Dictionary<string, Type>(StringComparer.OrdinalIgnoreCase);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeature.cs]: Hmmm, don't we use semantics equivalent to the current file system for view lookup? Wouldn't doing IgnoreCase here potentially break file systems that are case sensitive?

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs [hunk 4]
```diff
@@ -100,24 +69,15 @@ public CompilationResult Compile(RelativeFileInfo fileInfo, string compilationCo
...
-                options: _compilationOptions,
-                syntaxTrees: new[] { syntaxTree },
-                references: CompilationReferences);
-
-            compilation = Rewrite(compilation);
+            var syntaxTree = _compiler.CreateSyntaxTree(sourceText).WithFilePath(assemblyName);
+            var compilation = _compiler
+                .CreateCompilation(assemblyName)
+                .AddSyntaxTrees(syntaxTree);
+            compilation = ExpressionRewriter.Rewrite(compilation);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs]: Why break ExpressionRewriting out of the normal rewrite call?

Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs]: We have to invoke it from precompilation. I'll remove the method from here, or just leave it as part of the method but call the new static method instead.

File: src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeature.cs (hunk 3 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeature.cs]: We use case insensitive lookups in other view lookup places (for instance the `CompilerCache`). Sadly the case sensitivity here wouldn't make a big difference.

File: src/Microsoft.AspNetCore.Mvc.Core/IDesignTimeMvcBuilderConfiguration.cs (hunk 1 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/IDesignTimeMvcBuilderConfiguration.cs]: Sure, let's figure out the naming bit in a second iteration.

File: src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ViewInfoContainer.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ViewInfoContainer.cs]: Yup.

File: src/Microsoft.AspNetCore.Mvc.Core/IDesignTimeMvcBuilderConfiguration.cs (hunk 1 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/IDesignTimeMvcBuilderConfiguration.cs]: cc @rynowak \ @Eilon - suggestions?

**Pull request closed without merging**